### PR TITLE
Better record and value formatting in tools

### LINKED ISSUE #7019: Format records to look like canonical declarations
If I define a record where labels are on a new line:

```fsharp
type R =
    { MyInt: int
      MyString: string
      MyDateTime: System.DateTime }

type R' = { I: int }
```

It formats the tooltip with semicolons at the end of the labels:

image

Ditto for a single-label record:

image

This asymmetry is a bit bothersome.

### PR DESCRIPTION
**Pull request by cartermp**
Fixes #7019, #7163

Example code:

```fsharp
open System

type R =
    { MyInt: int
      MyString: string
      MyDateTime: DateTime }

type R' = { I: int }

type AR = {| MyInt: int; MyString: string; MyDateTime: DateTime |}

let ar = {| I = 12; Name = "phillip" |}
```

#### Records

Multiple labels:

image

Single label:

image

#### Anonymous records

Type alias (lol):

image

Expression with multiple labels:

image

Expression with single label:

image

**FSI formatting:**

image

- [x] Remove semicolons from formatting for Records
- [x] Spaces between braces for Records
- [x] Adjust semicolons and colons for Anonymous Records
- [x] Spaces between braces for Anonymous Records
- [x] Spaces between braces and removed semicolons for values in FSI
- [x] Place colon directly after value name in FSI
- [x] Update all tests

### CODE REVIEW

**Comment by auduchinok** [comment]
Could we also have a single space between braces and fields as it's usually formatted currently?

**Comment by smoothdeveloper** [comment]
Is there going to be other changes to harmonize what the "canonical" representation of the types looks like in the tooltips?

**Comment by cartermp** [comment]
@auduchinok Done:

image

(also for anonymous records)

@smoothdeveloper Maybe. I'm just doing this for records right now though.

**Comment by cartermp** [comment]
Ack, there's a missing space on the remaining records. Will adjust.

**Comment by cartermp** [comment]
So this is _almost_ good to go. If you zoom in, the top label as slightly less indentation than the rest. Can't quite figure out why.

File: vsintegration/tests/UnitTests/LegacyLanguageService/Tests.LanguageService.QuickInfo.fs [hunk 1]
```diff
@@ -762,7 +762,7 @@ Full name: Microsoft.FSharp.Control.Async""".TrimStart().Replace("\r\n", "\n")
       
         let fileContents = """namespace NS
                            type Re(*MarkerRecord*) = { X : int } """
-        let expectedQuickinfoTypeRecored = "type Re =  {X: int;}"
```
Comment (TIHan) [inline vsintegration/tests/UnitTests/LegacyLanguageService/Tests.LanguageService.QuickInfo.fs]: Yea, format before was, yuck.

**Comment by cartermp** [comment]
Note: also need to do anonymous records

**Comment by dsyme** [comment]
Thank you, this is a great change.  Little things like this area really important!

**Comment by cartermp** [comment]
Just need to figure out how to make this dang fsharpqa test pass; there's nothing in the log (@forki pointed this out here: https://github.com/dotnet/fsharp/issues/7066)

@dsyme / @KevinRansom since I've zero familiarity with this old test suite, would any of you have an idea?

**Comment by dsyme** [comment]
This is the failing test:
```
Printing (BindingsWithValues01.fsx) -- failed
```
That might be enough to track it down?

**Comment by forki** [comment]
The build is writing down which test fails. But we can't see the reason. So we have to try to reproduce locally. Which can be really hard. It would be really useful if the failed assertion would be printed as well.

**Comment by cartermp** [comment]
Yep, as @forki said. Especially with this it's a game of whack-a mole, since I already modified everything that I thought was relevant in that test, but it still failed.

**Comment by dsyme** [comment]
> The build is writing down which test fails. But we can't see the reason. So we have to try to reproduce locally. Which can be really hard. It would be really useful if the failed assertion would be printed as well.

Yeah I know, it sucks.  

I tend to run the commands implied by the `Env.lst` by hand when this happens

**Comment by dsyme** [comment]
Based on this:

	SOURCE=BindingsWithValues01.fsx COMPILE_ONLY=1 SCFLAGS="--nologo" FSIMODE=PIPE		# BindingsWithValues01.fsx

It would be something like

	fsi < BindingsWithValues01.fsx 

And I'd expect that to fail

**Comment by cartermp** [comment]
Found the bug!
image

**Comment by cartermp** [comment]
@dsyme quick questions: we currently format records in two places:

https://github.com/dotnet/fsharp/blob/77b552118f29fed68d18b447a74ba0ae25ab8b5c/src/fsharp/NicePrint.fs#L1740-L1743

and

https://github.com/dotnet/fsharp/blob/77b552118f29fed68d18b447a74ba0ae25ab8b5c/src/utils/sformat.fs#L762-L773

Any reason why this is split in these two places? The latter is used seemingly only by FSI when printing out the value of something.

**Comment by cartermp** [comment]
Okay, I'm going to ignore value printing for now and file a separate issue, as that all goes through separate code paths

**Comment by cartermp** [comment]
All green. @TIHan @KevinRansom PTAL

**Comment by dsyme** [comment]
> Any reason why this is split in these two places? The latter is used seemingly only by FSI when printing out the value of something.

One is incorporated into FSharp.Core - the other is in FSharp.Compiler.

**Comment by cartermp** [comment]
I'll also fix #7163  in here since it's easy enough

**Comment by dsyme** [comment]
Another one I noticed was that `val x : int` in output has a space between `x` and `:`, when we normally don't have that space in F# formatting conventions.

**Comment by cartermp** [comment]
I can git that 'uhn

**Comment by cartermp** [comment]
https://github.com/dotnet/fsharp/pull/7196#issuecomment-510529427

**Comment by cartermp** [comment]
omg it's green @TIHan @KevinRansom

**Pull request merged**